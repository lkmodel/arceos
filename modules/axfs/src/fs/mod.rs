cfg_if::cfg_if! {
    if #[cfg(feature = "myfs")] {
        pub mod myfs;
        pub const BLOCK_SIZE: usize = 512;
    } else if #[cfg(feature = "fatfs")] {
        pub mod fatfs;
        pub use fatfs::BLOCK_SIZE;
    } else if #[cfg(feature = "ext4fs")] {
        pub mod ext4fs;
        pub use ext4fs::BLOCK_SIZE;
    } else if #[cfg(feature = "ext4_rs")] {

        pub mod ext4;
        pub use ext4::BLOCK_SIZE;
    }
}

// #[cfg(feature = "ext4_rs")]
// pub mod ext4;

#[cfg(feature = "devfs")]
pub use axfs_devfs as devfs;

#[cfg(feature = "ramfs")]
pub use axfs_ramfs as ramfs;
