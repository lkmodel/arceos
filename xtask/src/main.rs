use clap::Parser;
use serde::Deserialize;
use std::fs::{File, OpenOptions};
use std::io::{Read, Seek, SeekFrom, Write};
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::str::from_utf8;
use std::{env, fs, io};

const CC: &str = "riscv64-linux-musl-gcc";
const DYNAMIC_FLAG: [&str; 1] = ["-fPIE", ];
const STATIC_FLAG: [&str; 0] = [];

#[derive(Parser, Debug)]
#[command(version, about, long_about = None)]
struct Args {
    /// Architecture to use (e.g., riscv64)
    #[arg(long, default_value = "riscv64")]
    arch: String,

    /// Log level (e.g., warn)
    #[arg(long, short, default_value = "warn")]
    log: String,

    /// Enable QEMU logging (y/n)
    #[arg(long, default_value = "n")]
    qemu_log: String,

    /// App type to run (all, static, dynamic)
    #[arg(long, short)]
    ttype: Option<String>,

    /// Review snapshot and don`t run app
    #[arg(long, short)]
    snapshot: bool,

    /// Which app to run
    app: String,
}

#[derive(Debug, Deserialize)]
struct Config {
    dev: DevConfig,
}

#[derive(Debug, Deserialize)]
struct DevConfig {
    rename: Option<String>,
    ttype: Option<String>,
    dynamic_flags: Option<Vec<String>>,
    static_flags: Option<Vec<String>>,
    snapshot: Option<bool>,
}

fn main() {
    let args = Args::parse();

    if args.snapshot {
        Command::new("cargo")
            .args(&["install", "cargo-insta"])
            .status()
            .expect("Can`t install cargo-insta");
        let current_dir = env::current_dir().expect("Failed to get current directory");
        Command::new("cargo")
            .args(&[
                "insta",
                "review",
                "--workspace-root",
                current_dir.join("payload").join(args.app).to_str().unwrap(),
            ])
            .status()
            .expect("Review failed");
        return;
    }

    let current_dir = env::current_dir().expect("Failed to get current directory");

    // Check and install musl-riscv64
    if !check_installation() {
        if !install_musl_riscv64() {
            eprintln!("Failed to install musl-riscv64");
            return;
        }
    }

    // Checkout the mocklibc branch
    if !check_branch("mocklibc") {
        eprintln!("Failed to switch to mocklibc branch");
        return;
    }

    // Add Rust target
    let status = Command::new("rustup")
        .args(["target", "add", "riscv64gc-unknown-linux-musl"])
        .status()
        .expect("Failed to add Rust target");

    if !status.success() {
        eprintln!("Failed to add Rust target");
        return;
    }

    let is_ci = env::var("CI").is_err();
    if args.app == "all".to_string() || !is_ci {
        traverse_all_app(&current_dir.join("payload"), &|dir: &PathBuf| -> Result<(), String> {
            let arg = Args {
                arch: "riscv64".to_string(),
                log: "warn".to_string(),
                qemu_log: "n".to_string(),
                ttype: None,
                snapshot: false,
                app: dir.iter().last().unwrap().to_str().unwrap().to_string(),
            };
            let config = parse_toml(&current_dir.join("payload").join(&arg.app));
            println!("APP:{}", arg.app);
            dynamic_test(&arg, &current_dir, &config)
        }).expect("Failed to traverse app")
            .iter()
            .for_each(|app| println!("{}\t{}", app[0], app[1]));
        return;
    }

    // Parse the toml file
    let config = parse_toml(&current_dir.join("payload").join(&args.app));
    println!("Config: {:?}", config);
    let mut ttype = config
        .as_ref()
        .and_then(|c| c.dev.ttype.as_ref())
        .unwrap()
        .as_str();
    let args_ttype: &String;
    if args.ttype.is_some() {
        args_ttype = args.ttype.as_ref().unwrap();
        if args_ttype == ttype || ttype == "all" {
            ttype = args_ttype;
        } else {
            eprintln!("Unsupported ttype {}", args_ttype);
            return;
        }
    }

    println!("Architecture: {}", args.arch);
    println!("Log level: {}", args.log);
    println!("QEMU log: {}", args.qemu_log);
    println!("Link type: {}", ttype);
    println!("App: {}", args.app);

    // Build mocklibc
    // let status = Command::new("cargo")
    //     .args([
    //         "build",
    //         "--target",
    //         "riscv64gc-unknown-linux-musl",
    //         "--release",
    //         "-p",
    //         "mocklibc",
    //     ])
    //     .status()
    //     .expect("Failed to build mocklibc");
    //
    // if !status.success() {
    //     eprintln!("Failed to build mocklibc");
    //     return;
    // }

    // Move the built library to the payload directory
    // let lib_path = PathBuf::from("./target/riscv64gc-unknown-linux-musl/release/libmocklibc.so");
    // let payload_path = PathBuf::from(format!("./payload/{}/libmocklibc.so",args.app));
    // std::fs::rename(&lib_path, &payload_path).expect("Failed to move libmocklibc.so");


    // Run tests based on the test type
    match ttype {
        "dynamic" => dynamic_test(&args, &current_dir, &config).unwrap(),
        "static" => static_test(&args, &current_dir, &config).unwrap(),
        "all" => {
            dynamic_test(&args, &current_dir, &config).unwrap();
            println!("----------------------------------------");
            println!(" ");
            println!("----------------------------------------");
            static_test(&args, &current_dir, &config).unwrap();
        }
        _ => eprintln!("Invalid test type"),
    }
}
fn traverse_all_app(path: &PathBuf, cb: &dyn Fn(&PathBuf) -> Result<(), String>) -> io::Result<Vec<Vec<String>>> {
    let mut apps = Vec::new();
    if path.is_dir() {
        for entry in fs::read_dir(path)? {
            let entry = entry?;
            let path = entry.path();
            if path.is_dir() {
                if let Err(e) = cb(&path) {
                    apps.push(vec![path.iter().last().unwrap().to_str().unwrap().to_string(), e]);
                } else {
                    apps.push(vec![path.iter().last().unwrap().to_str().unwrap().to_string(), String::from("OK")]);
                }
            }
        }
    }
    Ok(apps)
}

fn parse_toml(app_path: &PathBuf) -> Option<Config> {
    let toml_path = app_path.join("config.toml");
    if toml_path.exists() {
        let toml_content = fs::read_to_string(toml_path).expect("Failed to read toml file");
        let config: Config = toml::from_str(&toml_content).expect("Failed to parse toml file");
        Some(config)
    } else {
        None
    }
}

fn check_installation() -> bool {
    Command::new("which")
        .arg("riscv64-linux-musl-gcc")
        .output()
        .map(|output| output.status.success())
        .unwrap_or(false)
        || PathBuf::from("/opt/musl_riscv64").exists()
}

fn install_musl_riscv64() -> bool {
    if check_installation() {
        return true;
    }

    if !get_sudo() {
        return false;
    }

    let status = Command::new("git")
        .args(["clone", "https://github.com/richfelker/musl-cross-make.git"])
        .status()
        .expect("Failed to clone musl-cross-make");

    if !status.success() {
        eprintln!("Failed to clone musl-cross-make");
        return false;
    }

    let mut config_mak = std::fs::read_to_string("musl-cross-make/config.mak.dist")
        .expect("Failed to read config.mak.dist");

    config_mak.insert_str(15, "TARGET = riscv64-linux-musl\n");
    config_mak.insert_str(22, "OUTPUT = /opt/musl_riscv64\n");

    std::fs::write("musl-cross-make/config.mak", config_mak).expect("Failed to write config.mak");

    let status = Command::new("make")
        .args(["-j4"])
        .current_dir("musl-cross-make")
        .status()
        .expect("Failed to build musl-cross-make");

    if !status.success() {
        eprintln!("Build failed");
        return false;
    }

    let status = Command::new("sudo")
        .args(["make", "install", "-j4"])
        .current_dir("musl-cross-make")
        .status()
        .expect("Failed to install musl-cross-make");

    if !status.success() {
        eprintln!("Installation failed");
        return false;
    }

    // Add to PATH in .bashrc and .zshrc
    let home_dir = env::var("HOME").expect("Failed to get HOME directory");
    let bashrc_path = PathBuf::from(&home_dir).join(".bashrc");
    let zshrc_path = PathBuf::from(&home_dir).join(".zshrc");

    for rc_file in &[bashrc_path, zshrc_path] {
        if rc_file.exists() {
            let mut rc_content = std::fs::read_to_string(rc_file).expect("Failed to read rc file");
            if !rc_content.contains("/opt/musl_riscv64/bin") {
                rc_content.push_str("\nexport PATH=$PATH:/opt/musl_riscv64/bin\n");
                std::fs::write(rc_file, rc_content).expect("Failed to write rc file");
            }
        }
    }

    println!("Musl RISC-V64 toolchain installation complete");
    std::fs::remove_dir_all("musl-cross-make").expect("Failed to remove musl-cross-make directory");

    // Add musl-riscv64 to PATH
    let mut path = env::var("PATH").unwrap_or_default();
    path.push_str(":/opt/musl_riscv64/bin");
    unsafe {
        env::set_var("PATH", &path);
    }

    true
}

fn get_sudo() -> bool {
    for attempt in 1..=3 {
        println!("Requesting sudo permissions (attempt {}/3)", attempt);
        let status = Command::new("sudo")
            .arg("-v")
            .status()
            .expect("Failed to request sudo permissions");

        if status.success() {
            println!("Sudo permissions granted");
            return true;
        } else {
            println!("Permission denied, please retry");
        }
    }

    eprintln!("Failed to obtain sudo permissions after maximum attempts");
    false
}

fn check_branch(branch_name: &str) -> bool {
    let output = Command::new("git")
        .args(["rev-parse", "--abbrev-ref", "HEAD"])
        .output()
        .expect("Failed to get current branch");

    if output.status.success() {
        let current_branch = String::from_utf8(output.stdout).expect("Failed to parse branch name");
        if current_branch.trim() == branch_name {
            println!("Currently on branch {}", branch_name);
            true
        } else {
            println!("Not on branch {}", branch_name);
            let status = Command::new("git")
                .args(["switch", branch_name])
                .status()
                .expect("Failed to switch branch");

            if status.success() {
                println!("Switched to branch {}", branch_name);
                true
            } else {
                eprintln!("Failed to switch branch");
                false
            }
        }
    } else {
        eprintln!("Failed to get current branch");
        false
    }
}

fn static_test(args: &Args, current_dir: &PathBuf, config: &Option<Config>) -> Result<(), String> {
    let payload_dir = current_dir.join("payload");
    let app_path = payload_dir.join(args.app.as_str());
    build(&app_path, false, config).expect("Failed to build dynamic test");
    run_make(args, current_dir)
}

fn dynamic_test(args: &Args, current_dir: &PathBuf, config: &Option<Config>) -> Result<(), String> {
    let payload_dir = current_dir.join("payload");
    let app_path = payload_dir.join(args.app.as_str());
    build(&app_path, true, config).expect("Failed to build dynamic test");
    run_make(args, current_dir)
}

fn run_make(args: &Args, current_dir: &PathBuf) -> Result<(), String> {
    let status = Command::new("make")
        .args(["defconfig", "ARCH=riscv64"])
        .current_dir(&current_dir)
        .status()
        .expect("Failed to run make defconfig");

    if !status.success() {
        return Err(String::from("make defconfig failed"));
    }

    let status = Command::new("make")
        .args([
            "A=examples/loader",
            &format!("ARCH={}", args.arch),
            &format!("LOG={}", args.log),
            &format!("QEMU_LOG={}", args.qemu_log),
            "run",
        ])
        .current_dir(&current_dir)
        .status()
        .expect("Failed to run make");

    if !status.success() {
        return Err(String::from("make run failed"));
    }
    Ok(())
}

/// Generate a binary file from the ELF file
fn generate_bin(elf_file: &str, working_dir: &PathBuf) -> io::Result<()> {
    let elf_path = working_dir.join(elf_file);
    let bin_path = working_dir.join("apps.bin");

    // Create a 32MB empty file
    let file = File::create(&bin_path)?;
    file.set_len(32 * 1024 * 1024)?; // 32MB

    // Get the size of the ELF file
    let app_size = fs::metadata(&elf_path)?.len();

    // Convert the size to a 16-byte hex string and reverse the byte order
    let size_hex = format!("{:016x}", app_size);
    let mut size_bytes = hex::decode(size_hex).unwrap();
    size_bytes.reverse();

    // Write the size to the binary file (first 8 bytes)
    let mut file = OpenOptions::new().write(true).open(&bin_path)?;
    file.write_all(&size_bytes)?;

    // Write the ELF file content to the binary file (starting at offset 8)
    let mut elf_file = File::open(elf_path)?;
    let mut buffer = Vec::new();
    elf_file.read_to_end(&mut buffer)?;

    file.seek(SeekFrom::Start(8))?;
    file.write_all(&buffer)?;

    // Copy the binary file to the parent directory
    let parent_bin_path = working_dir.parent().unwrap().join("apps.bin");
    fs::rename(&bin_path, &parent_bin_path)?;
    Ok(())
}

/// Build a ELF file
fn build(elf_path: &PathBuf, ttype: bool, config: &Option<Config>) -> io::Result<()> {
    let elf_file = elf_path
        .iter()
        .last()
        .unwrap()
        .to_str()
        .unwrap()
        .split('_')
        .next()
        .unwrap()
        .to_string();
    // Determine the rename value from the config
    let rename = config
        .as_ref()
        .and_then(|c| c.dev.rename.as_ref())
        .unwrap_or(&elf_file)
        .as_str();
    let c_file = format!("{}.c", rename); // C source file
    let elf_output = format!("{}", rename); // Output ELF file

    // Determine the flags based on the config
    let dynamic_flags = config
        .as_ref()
        .and_then(|c| c.dev.dynamic_flags.as_ref())
        .map(|flags| flags.iter().map(|s| s.as_str()).collect::<Vec<_>>())
        .unwrap_or_else(|| DYNAMIC_FLAG.iter().map(|s| *s).collect());

    let static_flags = config
        .as_ref()
        .and_then(|c| c.dev.static_flags.as_ref())
        .map(|flags| flags.iter().map(|s| s.as_str()).collect::<Vec<_>>())
        .unwrap_or_else(|| STATIC_FLAG.iter().map(|s| *s).collect());

    // Compile the C file
    let mut status = Command::new(CC);
    let flags = if ttype {
        dynamic_flags
    } else {
        static_flags
    };
    status
        .args(&["-v", &c_file])
        .args(flags)
        .args(&["-o", &elf_output])
        .current_dir(elf_path)
        .stdout(Stdio::null())
        .stderr(Stdio::null());
    let re = status.status()?;

    if !re.success() {
        return Err(io::Error::new(
            io::ErrorKind::Other,
            "Failed to compile C file",
        ));
    }

    let mut insta_set = insta::Settings::clone_current();
    insta_set.set_snapshot_path(elf_path.join("snapshot"));
    insta_set.set_prepend_module_to_snapshot(false);

    let t_type: String;
    if ttype {
        t_type = "dynamic".to_string();
    } else {
        t_type = "static".to_string();
    }

    // Generate disassembly file
    let output = Command::new("riscv64-linux-musl-objdump")
        .args(&["-d", &elf_output])
        .current_dir(elf_path)
        .output()
        .expect("Failed to run riscv64-linux-musl-objdump");
    let output_file = elf_path.clone();

    if config
        .as_ref()
        .and_then(|c| c.dev.snapshot.as_ref())
        .is_some_and(|b| *b)
    {
        insta_set.set_snapshot_suffix(format!("{}_{}.S", elf_file, t_type));
        insta_set.bind(|| {
            insta::assert_snapshot!(from_utf8(&output.stdout).unwrap());
        });
    }
    fs::write(output_file.join(&format!("{}.S", elf_file)), output.stdout)
        .expect("Failed to write ELF file");

    // Generate ELF info file
    let output = Command::new("riscv64-linux-musl-readelf")
        .args(&["-a", &elf_output])
        .current_dir(elf_path)
        .output()
        .expect("Failed to run riscv64-linux-musl-readelf");
    let output_file = elf_path.clone();
    if config
        .as_ref()
        .and_then(|c| c.dev.snapshot.as_ref())
        .is_some_and(|b| *b)
    {
        insta_set.set_snapshot_suffix(format!("{}_{}.elf", elf_file, t_type));
        insta_set.bind(|| {
            insta::assert_snapshot!(from_utf8(&output.stdout).unwrap());
        });
    }
    fs::write(
        output_file.join(&format!("{}.elf", elf_file)),
        output.stdout,
    )
        .expect("Failed to write ELF file");

    // Generate full disassembly and symbol table
    let output = Command::new("riscv64-linux-musl-objdump")
        .args(&["-x", "-d", &elf_output])
        .current_dir(elf_path)
        .output()
        .expect("Failed to run riscv64-linux-musl-objdump");
    let output_file = elf_path.clone();
    if config
        .as_ref()
        .and_then(|c| c.dev.snapshot.as_ref())
        .is_some_and(|b| *b)
    {
        insta_set.set_snapshot_suffix(format!("{}_{}.dump", elf_file, t_type));
        insta_set.bind(|| {
            insta::assert_snapshot!(from_utf8(&output.stdout).unwrap());
        });
    }
    fs::write(
        output_file.join(&format!("{}.dump", elf_file)),
        output.stdout,
    )
        .expect("Failed to write ELF file");

    // Generate the binary file
    generate_bin(&elf_output, &elf_path)?;

    Ok(())
}

