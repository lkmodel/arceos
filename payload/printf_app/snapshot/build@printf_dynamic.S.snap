---
source: xtask/src/main.rs
expression: from_utf8(&output.stdout).unwrap()
---
printf:     file format elf64-littleriscv


Disassembly of section .plt:

00000000000005d0 <.plt>:
 5d0:	00002397          	auipc	t2,0x2
 5d4:	41c30333          	sub	t1,t1,t3
 5d8:	a383be03          	ld	t3,-1480(t2) # 2008 <__TMC_END__>
 5dc:	fd430313          	addi	t1,t1,-44
 5e0:	a3838293          	addi	t0,t2,-1480
 5e4:	00135313          	srli	t1,t1,0x1
 5e8:	0082b283          	ld	t0,8(t0)
 5ec:	000e0067          	jr	t3

00000000000005f0 <printf@plt>:
 5f0:	00002e17          	auipc	t3,0x2
 5f4:	a28e3e03          	ld	t3,-1496(t3) # 2018 <printf>
 5f8:	000e0367          	jalr	t1,t3
 5fc:	00000013          	nop

0000000000000600 <puts@plt>:
 600:	00002e17          	auipc	t3,0x2
 604:	a20e3e03          	ld	t3,-1504(t3) # 2020 <puts>
 608:	000e0367          	jalr	t1,t3
 60c:	00000013          	nop

0000000000000610 <sleep@plt>:
 610:	00002e17          	auipc	t3,0x2
 614:	a18e3e03          	ld	t3,-1512(t3) # 2028 <sleep>
 618:	000e0367          	jalr	t1,t3
 61c:	00000013          	nop

0000000000000620 <__libc_start_main@plt>:
 620:	00002e17          	auipc	t3,0x2
 624:	a10e3e03          	ld	t3,-1520(t3) # 2030 <__libc_start_main>
 628:	000e0367          	jalr	t1,t3
 62c:	00000013          	nop

Disassembly of section .text:

0000000000000630 <_start>:
 630:	00002197          	auipc	gp,0x2
 634:	1d018193          	addi	gp,gp,464 # 2800 <__global_pointer$>
 638:	850a                	mv	a0,sp
 63a:	00002597          	auipc	a1,0x2
 63e:	80658593          	addi	a1,a1,-2042 # 1e40 <_DYNAMIC>
 642:	ff017113          	andi	sp,sp,-16
 646:	a009                	j	648 <_start_c>

0000000000000648 <_start_c>:
 648:	410c                	lw	a1,0(a0)
 64a:	00850613          	addi	a2,a0,8
 64e:	4781                	li	a5,0
 650:	00002717          	auipc	a4,0x2
 654:	a2073703          	ld	a4,-1504(a4) # 2070 <_fini>
 658:	00002697          	auipc	a3,0x2
 65c:	9f06b683          	ld	a3,-1552(a3) # 2048 <_init>
 660:	00002517          	auipc	a0,0x2
 664:	a0853503          	ld	a0,-1528(a0) # 2068 <_GLOBAL_OFFSET_TABLE_+0x30>
 668:	bf65                	j	620 <__libc_start_main@plt>

000000000000066a <deregister_tm_clones>:
 66a:	00002517          	auipc	a0,0x2
 66e:	99e50513          	addi	a0,a0,-1634 # 2008 <__TMC_END__>
 672:	00002797          	auipc	a5,0x2
 676:	99678793          	addi	a5,a5,-1642 # 2008 <__TMC_END__>
 67a:	00a78863          	beq	a5,a0,68a <deregister_tm_clones+0x20>
 67e:	00002797          	auipc	a5,0x2
 682:	9e27b783          	ld	a5,-1566(a5) # 2060 <_ITM_deregisterTMCloneTable>
 686:	c391                	beqz	a5,68a <deregister_tm_clones+0x20>
 688:	8782                	jr	a5
 68a:	8082                	ret

000000000000068c <register_tm_clones>:
 68c:	00002517          	auipc	a0,0x2
 690:	97c50513          	addi	a0,a0,-1668 # 2008 <__TMC_END__>
 694:	00002597          	auipc	a1,0x2
 698:	97458593          	addi	a1,a1,-1676 # 2008 <__TMC_END__>
 69c:	8d89                	sub	a1,a1,a0
 69e:	4035d793          	srai	a5,a1,0x3
 6a2:	91fd                	srli	a1,a1,0x3f
 6a4:	95be                	add	a1,a1,a5
 6a6:	8585                	srai	a1,a1,0x1
 6a8:	c599                	beqz	a1,6b6 <register_tm_clones+0x2a>
 6aa:	00002797          	auipc	a5,0x2
 6ae:	9ae7b783          	ld	a5,-1618(a5) # 2058 <_ITM_registerTMCloneTable>
 6b2:	c391                	beqz	a5,6b6 <register_tm_clones+0x2a>
 6b4:	8782                	jr	a5
 6b6:	8082                	ret

00000000000006b8 <__do_global_dtors_aux>:
 6b8:	1141                	addi	sp,sp,-16
 6ba:	e022                	sd	s0,0(sp)
 6bc:	00002417          	auipc	s0,0x2
 6c0:	9c440413          	addi	s0,s0,-1596 # 2080 <completed.1>
 6c4:	00044783          	lbu	a5,0(s0)
 6c8:	e406                	sd	ra,8(sp)
 6ca:	eb95                	bnez	a5,6fe <__do_global_dtors_aux+0x46>
 6cc:	00002797          	auipc	a5,0x2
 6d0:	9747b783          	ld	a5,-1676(a5) # 2040 <__cxa_finalize>
 6d4:	c791                	beqz	a5,6e0 <__do_global_dtors_aux+0x28>
 6d6:	00002517          	auipc	a0,0x2
 6da:	92a53503          	ld	a0,-1750(a0) # 2000 <__dso_handle>
 6de:	9782                	jalr	a5
 6e0:	f8bff0ef          	jal	ra,66a <deregister_tm_clones>
 6e4:	00002797          	auipc	a5,0x2
 6e8:	96c7b783          	ld	a5,-1684(a5) # 2050 <__deregister_frame_info>
 6ec:	c791                	beqz	a5,6f8 <__do_global_dtors_aux+0x40>
 6ee:	00000517          	auipc	a0,0x0
 6f2:	0a250513          	addi	a0,a0,162 # 790 <__FRAME_END__>
 6f6:	9782                	jalr	a5
 6f8:	4785                	li	a5,1
 6fa:	00f40023          	sb	a5,0(s0)
 6fe:	60a2                	ld	ra,8(sp)
 700:	6402                	ld	s0,0(sp)
 702:	0141                	addi	sp,sp,16
 704:	8082                	ret

0000000000000706 <frame_dummy>:
 706:	00002797          	auipc	a5,0x2
 70a:	9727b783          	ld	a5,-1678(a5) # 2078 <__register_frame_info>
 70e:	cf99                	beqz	a5,72c <frame_dummy+0x26>
 710:	1141                	addi	sp,sp,-16
 712:	e406                	sd	ra,8(sp)
 714:	00002597          	auipc	a1,0x2
 718:	97458593          	addi	a1,a1,-1676 # 2088 <object.0>
 71c:	00000517          	auipc	a0,0x0
 720:	07450513          	addi	a0,a0,116 # 790 <__FRAME_END__>
 724:	9782                	jalr	a5
 726:	60a2                	ld	ra,8(sp)
 728:	0141                	addi	sp,sp,16
 72a:	b78d                	j	68c <register_tm_clones>
 72c:	b785                	j	68c <register_tm_clones>

000000000000072e <main>:
 72e:	1101                	addi	sp,sp,-32
 730:	ec06                	sd	ra,24(sp)
 732:	e822                	sd	s0,16(sp)
 734:	1000                	addi	s0,sp,32
 736:	00000517          	auipc	a0,0x0
 73a:	03a50513          	addi	a0,a0,58 # 770 <main+0x42>
 73e:	ec3ff0ef          	jal	ra,600 <puts@plt>
 742:	478d                	li	a5,3
 744:	fef42623          	sw	a5,-20(s0)
 748:	fec42783          	lw	a5,-20(s0)
 74c:	85be                	mv	a1,a5
 74e:	00000517          	auipc	a0,0x0
 752:	03250513          	addi	a0,a0,50 # 780 <main+0x52>
 756:	e9bff0ef          	jal	ra,5f0 <printf@plt>
 75a:	fec42783          	lw	a5,-20(s0)
 75e:	853e                	mv	a0,a5
 760:	eb1ff0ef          	jal	ra,610 <sleep@plt>
 764:	4781                	li	a5,0
 766:	853e                	mv	a0,a5
 768:	60e2                	ld	ra,24(sp)
 76a:	6442                	ld	s0,16(sp)
 76c:	6105                	addi	sp,sp,32
 76e:	8082                	ret
